libraries/abstractions/common_io/test/test_iot_sdio.c

```diff
@@ -0,0 +1,1042 @@
+                                   ucWriteData,
+                                   0xffff);
+    TEST_ASSERT_EQUAL(IOT_SDIO_INVALID_VALUE, lRetVal);
+
+    /* Closing a valid handle should return success. */
+    lRetVal = iot_sdio_close(xSdioHandle);
+    TEST_ASSERT_EQUAL(IOT_SDIO_SUCCESS, lRetVal);
+}
+
+
+/**
+ * @brief Test Function to test iot_sdio_ioctl().
+ *
+ */
+TEST( TEST_IOT_SDIO, AFQP_IotSdioIoctl )
+{
+    IotSdioSlotHandle_t xSdioHandle = NULL;
+    int32_t lRetVal;
+    uint8_t ucReadVal =0;
+    bool bCardPresent = false;
+    IotSdioPerFuncEnable_t xPerFuncEnable = {eSdioFunction0, true};
+    IoTSdioPowerOnOffCard_t xCardPowerCtl = {false, false};
+    IotSdioFunction_t eFunc = eSdioFunction0;
+
+    /* Open with valid host index and slot index should return a non NULL handle. */
+    xSdioHandle = iot_sdio_open(uctestIotSdioValidHostIdx,
+                                uctestIotSdioValidSlotIdx);
+    TEST_ASSERT_NOT_EQUAL( NULL, xSdioHandle );
+
+    lRetVal = iot_sdio_set_sdio_callback(xSdioHandle, xtestIotSdioTestCallback, NULL);
+    TEST_ASSERT_EQUAL(IOT_SDIO_SUCCESS, lRetVal);
+
+    lRetVal = iot_sdio_ioctl(xSdioHandle, eSDIOSetCardDetectParams, (void *)&xtestIotSdioCardDetectParam);
+    TEST_ASSERT_EQUAL(IOT_SDIO_SUCCESS, lRetVal);
+
+    xCardPowerCtl.bPowerOn = false;
+    lRetVal = iot_sdio_ioctl(xSdioHandle,
+                             eSDIOPowerOnOffCard,
+                             (void *)&xCardPowerCtl);
+    TEST_ASSERT_EQUAL(IOT_SDIO_SUCCESS, lRetVal);
+
+    lRetVal = iot_sdio_ioctl(xSdioHandle,
+                             eSDIOCheckCardPresence,
+                             (void *)&bCardPresent);
+    TEST_ASSERT_EQUAL(IOT_SDIO_SUCCESS, lRetVal);
+    TEST_ASSERT_EQUAL( true, bCardPresent);
+
+    xCardPowerCtl.bPowerOn = true;
+    lRetVal = iot_sdio_ioctl(xSdioHandle,
+                             eSDIOPowerOnOffCard,
+                             (void *)&xCardPowerCtl);
+    TEST_ASSERT_EQUAL(IOT_SDIO_SUCCESS, lRetVal);
+
+    lRetVal = iot_sdio_card_connect(xSdioHandle);
+    TEST_ASSERT_EQUAL(IOT_SDIO_SUCCESS, lRetVal);
+
+    /* Switch to high speed */
+    lRetVal = iot_sdio_ioctl(xSdioHandle,
+                             eSDIOSwitchToHighSpeed,
+                             NULL);
+    TEST_ASSERT_EQUAL(IOT_SDIO_SUCCESS, lRetVal);
+
+    /* Verify that EHS bit in Bus Speed Register is set */
+    lRetVal = iot_sdio_io_read_direct(xSdioHandle,
+                                eSdioFunction0,
+                                testIotSdio_EHS_ADDR,
+                                (uint8_t *)&ucReadVal);
+    TEST_ASSERT_EQUAL(IOT_SDIO_SUCCESS, lRetVal);
+    TEST_ASSERT_EQUAL(testIotSdio_EHS_MASK, testIotSdio_EHS_MASK&ucReadVal);
+
+    /* Set driver strength */
+    lRetVal = iot_sdio_ioctl(xSdioHandle,
+                             eSDIOSetDriverStrength,
+                             (void *)&etestIotSdioDriverType);
+    TEST_ASSERT_EQUAL(IOT_SDIO_SUCCESS, lRetVal);
```

Comment:
Is this a must to support? Can the implementation return NOT_SUPPORTED?

Reply:
Yes. Card side may selectively un-support some type. The following check is comprehensive to cover each case.